File: modules/bitchat-module/ios/BitChatNostrBridge.swift
import Foundation
import Combine
import ExpoModulesCore

/// Bridges BitChat's Nostr stack (NostrRelayManager, GeoRelayDirectory,
/// NostrIdentityBridge, NostrProtocol) to the Expo module event system.
///
/// Nostr used to live in JS (`features/bitchat/lib/BitChatService.ts`) with a
/// hardcoded relay list and the user's main Nostr pubkey signing every event.
/// This bridge gives us upstream's geohash-scoped relay selection (5 closest
/// via Haversine from a 304-entry CSV) and per-geohash identity derivation
/// (HMAC-SHA256 of device seed + geohash) for free.
///
/// `@MainActor` because upstream's NostrRelayManager and GeoRelayDirectory
/// are both `@MainActor`-isolated.
@MainActor
final class BitChatNostrBridge {
    static let shared = BitChatNostrBridge()

    private weak var module: BitChatModule?
    private let identityBridge = NostrIdentityBridge()
    private var relayManager: NostrRelayManager?
    private var currentGeohash: String?
    private var currentGeohashPubkey: String?
    /// Cached so we can decrypt inbound gift wraps without re-deriving on
    /// every event. Kept in sync with `currentGeohash` via joinGeohash /
    /// leaveGeohash.
    private var currentGeohashIdentity: NostrIdentity?
    private var currentGeohashRelays: [String] = []
    /// Dedup set for gift-wrap event IDs — NostrRelayManager may replay the
    /// same wrap from multiple relays.
    private var seenGiftWrapIDs = Set<String>()
    private var isStarted = false

    private init() {}

    /// Called from `OnCreate` (nonisolated). Hops to MainActor to store the
    /// weak module reference, so the rest of the class can stay MainActor-isolated.
    nonisolated func attach(module: BitChatModule) {
        Task { @MainActor in
            BitChatNostrBridge.shared.module = module
        }
    }

    // MARK: - Lifecycle

    func start() {
        guard !isStarted else { return }
        isStarted = true

        // NostrRelayManager.live() would pull in NetworkActivationService /
        // FavoritesPersistenceService / LocationChannelManager / TorManager —
        // none of which make sense for this app. Inject a minimal deps struct
        // that short-circuits every gate and uses a direct URLSession for WSS.
        let deps = NostrRelayManagerDependencies(
            activationAllowed: { true },
            userTorEnabled: { false },
            hasMutualFavorites: { false },
            // Report "authorized" so NostrRelayManager's default-relay policy
            // (hasMutualFavorites || hasLocationPermission) lets us use the
            // default relay pool for non-geohash traffic if it ever comes up.
            hasLocationPermission: { true },
            mutualFavoritesPublisher: Empty<Set<Data>, Never>(completeImmediately: false).eraseToAnyPublisher(),
            locationPermissionPublisher: Empty<LocationChannelManager.PermissionState, Never>(completeImmediately: false).eraseToAnyPublisher(),
            torEnforced: { false },
            torIsReady: { false },
            torIsForeground: { true },
            awaitTorReady: { completion in completion(false) },
            makeSession: { BridgeURLSessionAdapter(base: URLSession.shared) },
            scheduleAfter: { delay, action in
                DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: action)
            },
            now: Date.init
        )

        relayManager = NostrRelayManager(dependencies: deps)

        // GeoRelayDirectory.shared auto-inits via .live() and calls
        // loadLocalEntries() at init, which reads the bundled
        // online_relays_gps.csv (304 entries). Its remote refresh calls
        // TorManager.shared.awaitReady() which is stubbed to return false,
        // so the remote path silently no-ops — we just use the bundled CSV.
        _ = GeoRelayDirectory.shared
    }

    func stop() {
        leaveGeohash()
        relayManager?.disconnect()
        relayManager = nil
        isStarted = false
    }

    // MARK: - Geohash channel

    func joinGeohash(_ geohash: String) throws {
        guard isStarted, let relayManager else {
            throw BitChatNostrBridgeError.notStarted
        }

        // Unsubscribe any previous geohash subscription. Leave WebSockets warm.
        if let previous = currentGeohash {
            relayManager.unsubscribe(id: "geo-\(previous)")
            relayManager.unsubscribe(id: "geo-dm-\(previous)")
        }

        let identity = try identityBridge.deriveIdentity(forGeohash: geohash)
        currentGeohash = geohash
        currentGeohashIdentity = identity
        currentGeohashPubkey = identity.publicKeyHex
        currentGeohashRelays = GeoRelayDirectory.shared.closestRelays(
            toGeohash: geohash,
            count: TransportConfig.nostrGeoRelayCount
        )

        let since = Date().addingTimeInterval(-TransportConfig.nostrGeohashInitialLookbackSeconds)
        let filter = NostrFilter.geohashEphemeral(
            geohash,
            since: since,
            limit: TransportConfig.nostrGeohashInitialLimit
        )

        relayManager.subscribe(
            filter: filter,
            id: "geo-\(geohash)",
            relayUrls: currentGeohashRelays,
            handler: { [weak self] event in
                // Runs on MainActor (NostrRelayManager is @MainActor).
                self?.emit(event: event, forGeohash: geohash)
            },
            onEOSE: nil
        )

        // Subscribe to NIP-17 gift wraps (kind 1059) addressed to our
        // per-geohash pubkey. Upstream uses a 24h lookback for DMs; match that.
        let dmSince = Date().addingTimeInterval(-TransportConfig.nostrDMSubscribeLookbackSeconds)
        let dmFilter = NostrFilter.giftWrapsFor(pubkey: identity.publicKeyHex, since: dmSince)
        relayManager.subscribe(
            filter: dmFilter,
            id: "geo-dm-\(geohash)",
            relayUrls: currentGeohashRelays,
            handler: { [weak self] giftWrap in
                self?.handleGiftWrap(giftWrap, forGeohash: geohash)
            },
            onEOSE: nil
        )
    }

    func leaveGeohash() {
        guard let geohash = currentGeohash else { return }
        relayManager?.unsubscribe(id: "geo-\(geohash)")
        relayManager?.unsubscribe(id: "geo-dm-\(geohash)")
        currentGeohash = nil
        currentGeohashPubkey = nil
        currentGeohashIdentity = nil
        currentGeohashRelays = []
        seenGiftWrapIDs.removeAll()
    }

    // MARK: - Send

    func sendMessage(_ content: String, nickname: String?) throws {
        guard let relayManager else {
            throw BitChatNostrBridgeError.notStarted
        }
        guard let geohash = currentGeohash else {
            throw BitChatNostrBridgeError.noActiveGeohash
        }

        let identity = try identityBridge.deriveIdentity(forGeohash: geohash)
        let event = try NostrProtocol.createEphemeralGeohashEvent(
            content: content,
            geohash: geohash,
            senderIdentity: identity,
            nickname: nickname,
            teleported: false
        )
        relayManager.sendEvent(event, to: currentGeohashRelays)
    }

    // MARK: - Private (1:1) geohash DMs (NIP-17 gift wrap)

    /// Send an encrypted DM to another geohash participant addressed by their
    /// Nostr hex pubkey (the per-geohash-derived pubkey we saw on an inbound
    /// public event's `senderPubkey` field). Mirrors upstream
    /// NostrTransport.sendPrivateMessageGeohash:235.
    ///
    /// Embeds a `BitchatPacket(privateMessage)` in a `bitchat1:`-prefixed
    /// base64url string (so cross-client interop with upstream bitchat works),
    /// then wraps that string in a NIP-17 gift wrap (kind 1059) sent to the
    /// same geohash-closest relay set we use for public messages.
    func sendPrivateMessage(to recipientHex: String, content: String) throws {
        guard let relayManager else {
            throw BitChatNostrBridgeError.notStarted
        }
        guard let identity = currentGeohashIdentity else {
            throw BitChatNostrBridgeError.noActiveGeohash
        }
        // Don't let the user DM themselves — upstream bails silently; we bail
        // loudly so the JS caller can show a meaningful error.
        if recipientHex.lowercased() == identity.publicKeyHex.lowercased() {
            throw BitChatNostrBridgeError.cannotDMSelf
        }

        let messageID = UUID().uuidString
        // We don't have a mesh-side myPeerID for geohash-only users, so stamp
        // the embedded BitchatPacket with an empty senderID. The authoritative
        // identity on the wire is the gift-wrap's outer Nostr signature.
        let senderPeerID = PeerID(str: "")

        guard let embedded = NostrEmbeddedBitChat.encodePMForNostrNoRecipient(
            content: content,
            messageID: messageID,
            senderPeerID: senderPeerID
        ) else {
            throw BitChatNostrBridgeError.embedFailed
        }

        let giftWrap = try NostrProtocol.createPrivateMessage(
            content: embedded,
            recipientPubkey: recipientHex,
            senderIdentity: identity
        )
        // Send on the same 5-relay geohash set as public messages. Keeps the
        // DM reachable wherever the recipient is subscribed.
        relayManager.sendEvent(giftWrap, to: currentGeohashRelays)
    }

    /// Gift-wrap inbound handler. Unwraps (NIP-17 decrypt) → parses the
    /// `bitchat1:` base64url-encoded BitchatPacket → decodes the inner TLV
    /// PrivateMessagePacket → emits `onNostrPrivateMessage` to JS.
    private func handleGiftWrap(_ giftWrap: NostrEvent, forGeohash geohash: String) {
        // Dedup — the subscription's multiple relays may each replay the
        // same wrap. Cap the set so long-running channels don't grow
        // unbounded; upstream uses a similar MessageDeduplicationService.
        guard !seenGiftWrapIDs.contains(giftWrap.id) else { return }
        seenGiftWrapIDs.insert(giftWrap.id)
        if seenGiftWrapIDs.count > 2000 {
            // Cheap trim — drop the oldest half.
            let kept = Array(seenGiftWrapIDs).suffix(1000)
            seenGiftWrapIDs = Set(kept)
        }

        guard let identity = currentGeohashIdentity,
              geohash == currentGeohash else { return }

        // NIP-17 double unwrap (gift wrap → seal → rumor).
        let decrypted: (content: String, senderPubkey: String, timestamp: Int)
        do {
            decrypted = try NostrProtocol.decryptPrivateMessage(
                giftWrap: giftWrap,
                recipientIdentity: identity
            )
        } catch {
            // Not addressed to us, or malformed — swallow.
            return
        }

        // The rumor's `content` is the bitchat1: base64url-encoded BitchatPacket.
        // Upstream's `NostrProtocol.base64URLDecode` is fileprivate so we
        // inline the equivalent 5-line transform here.
        guard decrypted.content.hasPrefix("bitchat1:") else { return }
        let encoded = String(decrypted.content.dropFirst("bitchat1:".count))
        guard let packetData = Self.base64URLDecode(encoded),
              let packet = BitchatPacket.from(packetData),
              packet.type == MessageType.noiseEncrypted.rawValue else {
            return
        }

        // Upstream's inner payload format: [1-byte NoisePayloadType][TLV].
        guard packet.payload.count >= 1,
              let payloadType = NoisePayloadType(rawValue: packet.payload[0]),
              payloadType == .privateMessage else {
            return
        }
        let tlv = packet.payload.subdata(in: 1..<packet.payload.count)
        guard let pm = PrivateMessagePacket.decode(from: tlv) else { return }

        module?.sendEvent("onNostrPrivateMessage", [
            "id": pm.messageID,
            "senderPubkey": decrypted.senderPubkey,
            "sender": String(decrypted.senderPubkey.prefix(12)),
            "content": pm.content,
            // NostrProtocol.decryptPrivateMessage returns the rumor timestamp
            // in Nostr seconds — convert to ms for JS consumers.
            "timestamp": Double(decrypted.timestamp) * 1000,
            "geohash": geohash,
            "isOwn": false,
        ])
    }

    // MARK: - Helpers

    /// Local copy of upstream's `NostrProtocol.base64URLDecode` (which is
    /// fileprivate so we can't call it directly from this module).
    /// Reverses base64url (`-_` alphabet, no padding) → standard base64 →
    /// bytes. Keeps the same byte output as upstream so inbound gift wraps
    /// from either client decode identically.
    private static func base64URLDecode(_ s: String) -> Data? {
        var str = s
        let pad = (4 - (str.count % 4)) % 4
        if pad > 0 { str += String(repeating: "=", count: pad) }
        str = str.replacingOccurrences(of: "-", with: "+")
                 .replacingOccurrences(of: "_", with: "/")
        return Data(base64Encoded: str)
    }

    // MARK: - Event dispatch

    private func emit(event: NostrEvent, forGeohash geohash: String) {
        // Extract #n (nickname) tag; fall back to short pubkey if absent.
        var nickname: String?
        for tag in event.tags where tag.count >= 2 {
            if tag[0] == "n" {
                nickname = tag[1]
                break
            }
        }
        let sender = nickname ?? String(event.pubkey.prefix(12))

        let payload: [String: Any] = [
            "id": event.id,
            "content": event.content,
            "sender": sender,
            "senderPubkey": event.pubkey,
            "timestamp": Double(event.created_at) * 1000,
            "geohash": geohash,
            "isOwn": event.pubkey == currentGeohashPubkey,
        ]
        module?.sendEvent("onNostrMessage", payload)
    }
}

// MARK: - Errors

enum BitChatNostrBridgeError: Error, LocalizedError {
    case notStarted
    case noActiveGeohash
    case cannotDMSelf
    case embedFailed

    var errorDescription: String? {
        switch self {
        case .notStarted: return "Nostr stack is not started. Call startNostr() first."
        case .noActiveGeohash: return "No active geohash. Call joinGeohash() first."
        case .cannotDMSelf: return "Cannot DM your own geohash identity."
        case .embedFailed: return "Failed to encode private message packet."
        }
    }
}

// MARK: - URLSession adapter for NostrRelayManager dependency injection

// Upstream's URLSessionAdapter / URLSessionWebSocketTaskAdapter are `private`
// to NostrRelayManager.swift, so we re-implement the same thin shape here.
private final class BridgeURLSessionWebSocketTaskAdapter: NostrRelayConnectionProtocol {
    private let base: URLSessionWebSocketTask

    init(base: URLSessionWebSocketTask) {
        self.base = base
    }

    func resume() {
        base.resume()
    }

    func cancel(with closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        base.cancel(with: closeCode, reason: reason)
    }

    func send(_ message: URLSessionWebSocketTask.Message, completionHandler: @escaping (Error?) -> Void) {
        base.send(message, completionHandler: completionHandler)
    }

    func receive(completionHandler: @escaping (Result<URLSessionWebSocketTask.Message, Error>) -> Void) {
        base.receive(completionHandler: completionHandler)
    }

    func sendPing(pongReceiveHandler: @escaping (Error?) -> Void) {
        base.sendPing(pongReceiveHandler: pongReceiveHandler)
    }
}

private struct BridgeURLSessionAdapter: NostrRelaySessionProtocol {
    let base: URLSession

    func webSocketTask(with url: URL) -> NostrRelayConnectionProtocol {
        BridgeURLSessionWebSocketTaskAdapter(base: base.webSocketTask(with: url))
    }
}

File: app/modules/bitchat-module/android/src/main/java/expo/modules/bitchat/BitChatNostrBridge.kt
package expo.modules.bitchat

import android.app.Application
import android.content.Context
import android.util.Base64
import com.bitchat.android.model.NoisePayloadType
import com.bitchat.android.model.PrivateMessagePacket
import com.bitchat.android.nostr.NostrEmbeddedBitChat
import com.bitchat.android.nostr.NostrEvent
import com.bitchat.android.nostr.NostrFilter
import com.bitchat.android.nostr.NostrIdentity
import com.bitchat.android.nostr.NostrIdentityBridge
import com.bitchat.android.nostr.NostrKind
import com.bitchat.android.nostr.NostrProtocol
import com.bitchat.android.nostr.NostrRelayManager
import com.bitchat.android.nostr.RelayDirectory
import com.bitchat.android.protocol.BitchatPacket
import com.bitchat.android.protocol.MessageType
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class BitChatNostrNotStartedException :
    Exception("Nostr stack is not started. Call startNostr() first.")

class BitChatNoActiveGeohashException :
    Exception("No active geohash. Call joinGeohash() first.")

class BitChatCannotDMSelfException : Exception("Cannot DM your own geohash identity.")

class BitChatEmbedFailedException : Exception("Failed to encode private message packet.")

/**
 * Bridges the vendored bitchat-android Nostr stack (NostrRelayManager,
 * RelayDirectory, NostrIdentityBridge, NostrProtocol) to the Expo module
 * event system — the Kotlin counterpart of ios/BitChatNostrBridge.swift with
 * identical event names and payload shapes.
 *
 * Simpler than iOS: the Android NostrRelayManager has geohash relay routing
 * built in (subscribeForGeohash / sendEventToGeohash + RelayDirectory), so no
 * dependency-injection dance is needed. The Tor-aware OkHttpProvider is
 * replaced by a plain stub (com.bitchat.android.net.OkHttpProvider in this
 * module's sources).
 *
 * Constants mirror upstream iOS TransportConfig: 1h public lookback /
 * limit 200, 24h gift-wrap lookback, 5 geohash relays.
 */
object BitChatNostrBridge {

    private const val PUBLIC_LOOKBACK_MS = 3_600_000L
    private const val PUBLIC_INITIAL_LIMIT = 200
    private const val DM_LOOKBACK_MS = 86_400_000L
    private const val GIFT_WRAP_SEEN_CAP = 2000
    private const val GIFT_WRAP_SEEN_KEEP = 1000

    var emitter: ((name: String, body: Map<String, Any?>) -> Unit)? = null

    private var appContext: Context? = null
    private var scopedContext: Context? = null
    private var activeScopeSuffix: String? = null
    private var isStarted = false
    private var currentGeohash: String? = null
    private var currentIdentity: NostrIdentity? = null
    private val seenGiftWrapIDs = LinkedHashSet<String>()
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    fun attach(context: Context, emit: (String, Map<String, Any?>) -> Unit) {
        appContext = context.applicationContext
        emitter = emit
    }

    fun detach() {
        stop()
        emitter = null
        appContext = null
    }

    fun start(profileScope: String) {
        val context = appContext ?: throw BitChatNostrNotStartedException()
        val suffix = BitchatProfileScope.storageSuffix(profileScope)
        if (isStarted && activeScopeSuffix == suffix) return
        if (isStarted) stop()

        (context.applicationContext as? Application)?.let { RelayDirectory.initialize(it) }
        // Per-profile device seed: NostrIdentityBridge reads/writes
        // `nostr_device_seed` through SecureIdentityStateManager(context), so a
        // scoped context isolates it per profile — same as the iOS
        // ProfileScopedBitchatKeychain. NOTE: NostrIdentityBridge keeps a
        // process-wide geohash→identity cache; Sovran profile switching is
        // restart-based, which keeps that cache scope-consistent.
        scopedContext = ProfileScopedContext(context, suffix)
        activeScopeSuffix = suffix
        isStarted = true
    }

    fun stop() {
        leaveGeohash()
        if (isStarted) {
            NostrRelayManager.shared.disconnect()
        }
        isStarted = false
        activeScopeSuffix = null
        scopedContext = null
    }

    // MARK: - Geohash channel

    fun joinGeohash(geohash: String) {
        if (!isStarted) throw BitChatNostrNotStartedException()
        val context = scopedContext ?: throw BitChatNostrNotStartedException()
        val relayManager = NostrRelayManager.shared

        currentGeohash?.let { previous ->
            relayManager.unsubscribe("geo-$previous")
            relayManager.unsubscribe("geo-dm-$previous")
        }

        val identity = NostrIdentityBridge.deriveIdentity(geohash, context)
        currentGeohash = geohash
        currentIdentity = identity

        val now = System.currentTimeMillis()
        relayManager.subscribeForGeohash(
            geohash = geohash,
            filter = NostrFilter.geohashEphemeral(
                geohash,
                since = now - PUBLIC_LOOKBACK_MS,
                limit = PUBLIC_INITIAL_LIMIT,
            ),
            id = "geo-$geohash",
            handler = { event -> emitPublicEvent(event, geohash) },
        )
        relayManager.subscribeForGeohash(
            geohash = geohash,
            filter = NostrFilter.giftWrapsFor(identity.publicKeyHex, since = now - DM_LOOKBACK_MS),
            id = "geo-dm-$geohash",
            handler = { giftWrap -> handleGiftWrap(giftWrap, geohash) },
        )
    }

    fun leaveGeohash() {
        val geohash = currentGeohash ?: return
        NostrRelayManager.shared.unsubscribe("geo-$geohash")
        NostrRelayManager.shared.unsubscribe("geo-dm-$geohash")
        currentGeohash = null
        currentIdentity = null
        synchronized(seenGiftWrapIDs) { seenGiftWrapIDs.clear() }
    }

    // MARK: - Send

    fun sendMessage(content: String, nickname: String?) {
        if (!isStarted) throw BitChatNostrNotStartedException()
        val geohash = currentGeohash ?: throw BitChatNoActiveGeohashException()
        val identity = currentIdentity ?: throw BitChatNoActiveGeohashException()
        // createEphemeralGeohashEvent is suspend (PoW mining support — Sovran
        // ships with PoW disabled, the default). Fire-and-forget like the iOS
        // bridge: the relay manager queues the actual socket write either way.
        scope.launch {
            val event = NostrProtocol.createEphemeralGeohashEvent(
                content = content,
                geohash = geohash,
                senderIdentity = identity,
                nickname = nickname?.takeIf { it.isNotEmpty() },
            )
            NostrRelayManager.shared.sendEventToGeohash(event, geohash)
        }
    }

    /**
     * NIP-17 gift-wrapped DM to another participant of the current geohash,
     * addressed by the per-geohash hex pubkey seen on their public messages.
     * The embedded `bitchat1:` BitchatPacket keeps cross-client interop with
     * upstream bitchat. Mirrors ios/BitChatNostrBridge.sendPrivateMessage.
     */
    fun sendPrivateMessage(recipientPubkey: String, content: String) {
        if (!isStarted) throw BitChatNostrNotStartedException()
        val geohash = currentGeohash ?: throw BitChatNoActiveGeohashException()
        val identity = currentIdentity ?: throw BitChatNoActiveGeohashException()
        if (recipientPubkey.equals(identity.publicKeyHex, ignoreCase = true)) {
            throw BitChatCannotDMSelfException()
        }

        val messageID = UUID.randomUUID().toString()
        // No mesh-side peerID for geohash-only users — empty senderID, same as
        // iOS. The authoritative identity is the gift wrap's outer signature.
        val embedded = NostrEmbeddedBitChat.encodePMForNostrNoRecipient(
            content = content,
            messageID = messageID,
            senderPeerID = "",
        ) ?: throw BitChatEmbedFailedException()

        val giftWraps = NostrProtocol.createPrivateMessage(
            content = embedded,
            recipientPubkey = recipientPubkey,
            senderIdentity = identity,
        )
        for (wrap in giftWraps) {
            NostrRelayManager.shared.sendEventToGeohash(wrap, geohash)
        }
    }

    // MARK: - Inbound

    private fun emitPublicEvent(event: NostrEvent, geohash: String) {
        // The geohash filter also matches kind-20001 presence heartbeats; only
        // kind 20000 is chat (same split as the vendor's GeohashMessageHandler).
        if (event.kind != NostrKind.EPHEMERAL_EVENT) return
        val nickname = event.tags.firstOrNull { it.size >= 2 && it[0] == "n" }?.get(1)
        emitter?.invoke(
            "onNostrMessage",
            mapOf(
                "id" to event.id,
                "content" to event.content,
                "sender" to (nickname ?: event.pubkey.take(12)),
                "senderPubkey" to event.pubkey,
                "timestamp" to event.createdAt.toDouble() * 1000,
                "geohash" to geohash,
                "isOwn" to event.pubkey.equals(currentIdentity?.publicKeyHex, ignoreCase = true),
            ),
        )
    }

    private fun handleGiftWrap(giftWrap: NostrEvent, geohash: String) {
        synchronized(seenGiftWrapIDs) {
            if (!seenGiftWrapIDs.add(giftWrap.id)) return
            if (seenGiftWrapIDs.size > GIFT_WRAP_SEEN_CAP) {
                val keep = seenGiftWrapIDs.toList().takeLast(GIFT_WRAP_SEEN_KEEP)
                seenGiftWrapIDs.clear()
                seenGiftWrapIDs.addAll(keep)
            }
        }
        val identity = currentIdentity ?: return
        if (geohash != currentGeohash) return

        // NIP-17 double unwrap (gift wrap → seal → rumor); not-for-us or
        // malformed wraps return null and are swallowed.
        val (content, senderPubkey, timestampSeconds) =
            NostrProtocol.decryptPrivateMessage(giftWrap, identity) ?: return

        if (!content.startsWith("bitchat1:")) return
        val packetData = base64URLDecode(content.removePrefix("bitchat1:")) ?: return
        val packet = BitchatPacket.fromBinaryData(packetData) ?: return
        if (packet.type != MessageType.NOISE_ENCRYPTED.value) return

        // Inner payload format: [1-byte NoisePayloadType][TLV].
        val payload = packet.payload
        if (payload.isEmpty()) return
        if (NoisePayloadType.fromValue(payload[0].toUByte()) != NoisePayloadType.PRIVATE_MESSAGE) return
        val pm = PrivateMessagePacket.decode(payload.copyOfRange(1, payload.size)) ?: return

        emitter?.invoke(
            "onNostrPrivateMessage",
            mapOf(
                "id" to pm.messageID,
                "senderPubkey" to senderPubkey,
                "sender" to senderPubkey.take(12),
                "content" to pm.content,
                "timestamp" to timestampSeconds.toDouble() * 1000,
                "geohash" to geohash,
                "isOwn" to false,
            ),
        )
    }

    /** base64url (`-_`, unpadded) → bytes; same transform as the iOS bridge. */
    private fun base64URLDecode(encoded: String): ByteArray? = try {
        Base64.decode(encoded, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    } catch (_: Exception) {
        null
    }
}
